### ISSUE TITLE
Refine the storage spec in HarborCluster CRD to provide more clear and flexible storage configurations

### ISSUE THREAD

**Issue by steven-zou** [goharbor/harbor-operator#642]
CRD version -> v1beta1?

```YAML
storage:
  kind: filesystem # or oss, swift, gcs, azure etc.
  spec:
    # different spec for different storage driver
    chartPersistentVolume: # Optional
        # Inline the corev1.PersistentVolumeClaimVolumeSource
        # ClaimName is the name of a PersistentVolumeClaim in the same namespace as the pod using this volume.
        # More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#persistentvolumeclaims
        claimName: myPVC # Required
        # Will force the ReadOnly setting in VolumeMounts.
        readOnly: false # Optional
        prefix: myPrefix # Optional
      registryPersistentVolume: # Optional
        # ... Skipped the same fields with 'chartPersistentVolume': 'claimName', 'readOnly' and 'prefix'.
        # ...
        # Max threads
        maxthreads: 100 # Optional, default = 100, minimal = 25
```

For the current incluster storage, use kind `minio`

```yaml
storage:
  kind: minio/minio
  spec:
      operatorVersion: 4.0.6
    # Image name for the MinIO. It will override the default one.
      image: my-minio # Optional
      # Image pull policy. It will override the global 'imageSource' settings and the default one.
      imagePullPolicy: # Optional, default = IfNotPresent
      # Image pull secrets. It will override the global 'imageSource' settings if it has been set.
      imagePullSecrets:
        - name: myHarborRegSecretOfMinIO
      # Redirection configurations.
      redirect: # Required
        # Determine if the redirection of minio storage is enabled.
        enable: true # Optional, default=true
        # If 'enable' is set to be true, then configure extra settings here.
        # Expose MinIO service for client accessing.
        # Same configuration with the top level 'expose' section.
        expose: # Optional
          # TLS setting
          tls: # Optional
            # Certificate reference
            certificateRef: <cert-ref> # Optional
          # Expose service with ingress way
          ingress:
            # Host of the exposed service
            host: <minio.goharbor.io> # Required
            # Ingress controller type, support ["gce","ncp","default"]
            controller: default # Optional, default value = "default"
            # Annotations applied to the ingress
            annotations: # Optional
              key: value
      # Reference to the secret containing the MinIO access key and secret key.
      secretRef: minioSecret # Optional
      # Supply number of replicas.
      # For standalone mode, supply 1. For distributed mode, supply 4 or more.
      # Note that the operator does not support upgrading from standalone to distributed mode.
      # Specially, 'replicas'*'volumesPerServer' should be >=4.
      replicas: 4 # Required, minimal=1
      # Number of persistent volumes that will be attached per server
      # Specially, 'replicas'*'volumesPerServer' should be >=4.
      volumesPerServer: 1 # Required, minimal=1
      # VolumeClaimTemplate allows a user to specify how volumes inside a MinIOInstance.
      # More info: https://github.com/kubernetes/kubernetes/blob/6b1d87acf3c8253c123756b9e61dac642678305f/pkg/apis/core/types.go#L398
      volumeClaimTemplate: {} # Optional
      # If provided, use these requests and limit for cpu/memory resource allocation
      # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
      resources: # Optional
        # Limits describes the maximum amount of compute resources allowed.
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        limits: {} # Optional
        # Requests describes the minimum amount of compute resources required.
        # If Requests is omitted for a container, it defaults to Limits if that is explicitly specified,
        # otherwise to an implementation-defined value.
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        requests: {} # Optional
```

**Comment by steven-zou** [comment]
@holyhope mentioned there is a RemoteObject, maybe we can do investigate and leverage that object here.

**Comment by holyhope** [comment]
Hello,

I was talking about the [`corev1.TypedLocalObjectReference{}`](https://pkg.go.dev/k8s.io/api/core/v1?utm_source=gopls#TypedLocalObjectReference) type.
Thanks to that type, we are able to define different CRD for each storage and have a single field to reference it in the `HabrorCluster` CRD.

wdyt?

**Comment by steven-zou** [comment]
> Hello,
> 
> I was talking about the [`corev1.TypedLocalObjectReference{}`](https://pkg.go.dev/k8s.io/api/core/v1?utm_source=gopls#TypedLocalObjectReference) type.
> Thanks to that type, we are able to define different CRD for each storage and have a single field to reference it in the `HabrorCluster` CRD.
> 
> wdyt?

Cool! Let me have a check.

**Comment by steven-zou** [comment]
@holyhope 

` corev1.TypedLocalObjectReference{}` is used to refer to another CR object which has been created in the same ns. If we adopt this way, that means the user needs to create other CRs before creating HarborCluster CR and will make the cluster creation more complicated then.

**Comment by holyhope** [comment]
Yes you are absolutely right.
This may be used at the Harbor, Registry and Chartmuseum resources levels.
But I agree, let's keep HarborCluster resource simple.
If I remember well, there is also a way to embed resources in other one, maybe this could be good?

**Comment by steven-zou** [comment]
> This may be used at the Harbor, Registry and Chartmuseum resources levels.

Correct! Do you want to introduce this fix at 1.1? or in future releases?

**Comment by steven-zou** [comment]
@holyhope 

> If I remember well, there is also a way to embed resources in other one, maybe this could be good?

If there is a better way, we can try. If there is not, then we can follow the current proposing way.
